---
title: "Assignment 3"
author: "Charlene Harasym, Jackson Chen, Kira Li"
date: "30/10/2021"
output: html_document
---


```{r}
#setwd(choose.dir())
```
Question 1.1 
```{r}
BAMFile<-read.csv("single_cell_RNA_seq_bam.sam", nrows=73, sep="\t", header=FALSE, fill=TRUE)
print(BAMFile)
```
SN tags indicate the reference sequence name. LN tags indicate the reference sequence length. 


Question 1.2
```{r}
SNXIndex<-which(BAMFile$V2=="SN:X")
XLength=BAMFile$V3[SNXIndex]
print("The length of chromosome X is 171031299 nucleotides")
```

Question 2.1
```{r}
sam <- read.csv("single_cell_RNA_seq_bam.sam", sep="\t", header=FALSE, comment.char="@", col.names = paste0("V",seq_len(30)), fill=TRUE)
sam <- sam[paste0("V",seq_len(11))]
print(paste(nrow(sam)))
print("There are 146346 reads in this BAM file")
```

Question 2.2 
```{r}
sam[10,]
```
Look at column V3 to find the chromosome to which the read was aligned. Column V11 contains the base quality score in ASCII. 

Question 2.3
```{r}
count=0
for(i in 1:nrow(sam)){
  if(sam$V3[i]=="X"){
    count=count+1
  }
}
print(paste(count,"reads align to chromosome X"))
```
Question 2.4
```{r}

qualityScores <- sam[which(sam$V3=="X"), 11]
#Scores <- qualityScores(filename=sam$V11,offset=33,nreads=1000)
#qualityScores <- paste0(XRows[,11])
qualityScoresUse <- paste0(qualityScores, collapse = "")
meanScore <- mean(utf8ToInt(qualityScoresUse))-33
print(paste(meanScore))
#qualityScores[1]
```
The mean base quality score for reads aligning to chromosome X is 32.72349

Question 2.5 
```{r}
qual <- c()
i <- 1
for(i in 1:5999) { 
  qual[i] <- list(strtoi(charToRaw(qualityScores[i]),16L)-33)
}
boxplot(qual, ylim=c(10,40),col="yellow",xlab="Read Position",ylab="Base Quality", outline = FALSE )
```
At this scale, it is difficult to see individual outliers in base quality reads, however, it can be observed at the average base quality read tended towards the low 30s. 


Question 2.6
Column V4

Question 2.7 
```{r}
count=0
for(i in 1:nrow(sam)){
  if(sam$V3[i]==9&sam$V4[i]>=40801273&sam$V4[i]<=40805199){
    count=count+1
  }
}

print(paste(count,"reads have their leftmost mapping position aligned with these coordinates"))

```
Question 2.8
```{r}
count=0
for(i in 1:nrow(sam)){
  if(sam$V5[i]<50){
    count=count+1
  }
}
print(paste(count, "reads have mapping quality less than 50"))
```
Question 2.9 
```{r}
count=0
total=0
for(i in 1:nrow(sam)){
  if(sam$V5[i]<50){
    count=count+1
    total=total+sam$V5[i]
  }
}

print(paste("the mean mapping quality for reads with mapping quality less than 50 is",total/count))
```

Question 2.10 
```{r}
count=0
for(i in 1:nrow(sam)){
  if(sam$V3[i]=="tdTomato"){
    count=count+1
  }
}
print(paste(count,"reads align to chromosome tdTomato"))
```
Since only 63 out of 146346 reads align to the fluorescence chromosome, the fluorescence would likely not be seen in the cell culture. Adding fluorescence to a genome could be used in order to see where within a cell certain genes are expressed. 

Question 3.1
```{r}
vcf_con <- file("RNA_seq_annotated_variants.vcf", open="r")
vcf_file <- readLines(vcf_con)
close(vcf_con)
vcf <- data.frame(vcf_file)
header <- vcf[grepl("##", vcf$vcf_file), ]
factor(header)
variants <- read.csv("RNA_seq_annotated_variants.vcf", skip=length(header), header=TRUE, sep="\t")

```
For the first variant, the reference allele is G and the alternate is A.

Question 3.2 
```{r}
line1string<-as.character(variants$INFO[1])
strings<-strsplit(line1string,';')

grep("ANN",strings[[1]], value=TRUE)



```
Question 3.3 
The annotation field tells us that the first variant is within an intron, does not have an impact, is from the Sulf1 gene that has a Gene ID of ENSMUSG00000016918, is a feature type "transcript" with Feature ID ENSMUST00000088585.9, it is protein coding, the ratio of exons to introns is 2:21.

Question 3.4 
```{r}
line1string<-as.character(variants$INFO[683])
strings<-strsplit(line1string,';')

grep("ANN",strings[[1]], value=TRUE)
```
This variant would affect the Rps19 gene. 

Question 3.5 
```{r}
print(paste("# Chromosome Number Variations:: ", length(grep("chromosome_number_variation", variants$INFO))))
print(paste("# Exon Loss Variants: ", length(grep("exon_loss_variant", variants$INFO))))
print(paste("# Frameshift Variants: ", length(grep("frameshift_variant", variants$INFO))))
print(paste("# Rare Amino Acid Variants: ", length(grep("rare_amino_acid_variant", variants$INFO))))
print(paste("# Splice Acceptor Variants: ", length(grep("splice_acceptor_variant", variants$INFO))))
print(paste("# Splice Donor Variants: ", length(grep("splice_donor_variant", variants$INFO))))
print(paste("# Start Lost: ", length(grep("start_lost", variants$INFO))))
print(paste("# Stop Gained: ", length(grep("stop_gained", variants$INFO))))
print(paste("# Stop Lost: ", length(grep("stop_lost", variants$INFO))))
print(paste("# Transcript Ablation: ", length(grep("transcript_ablation", variants$INFO))))
print(paste("3 Prime UTR Truncation & Exon Loss: ", length(grep("3_prime_UTR_truncation & exon_loss", variants$INFO))))
print(paste("5 Prime UTR Truncation & Exon Loss Variant: ", length(grep("5_prime_UTR_truncation & exon_loss_variant", variants$INFO))))
print(paste("Coding Sequence Variant: ", length(grep("coding_sequence_variant", variants$INFO))))
print(paste("Conservative Inframe Deletion: ", length(grep("conservative_inframe_deletion", variants$INFO))))
print(paste("Conservative Inframe Insertion: ", length(grep("conservative_inframe_insertion", variants$INFO))))
print(paste("Disruptive Inframe Deletion: ", length(grep("disruptive_inframe_deletion", variants$INFO))))
print(paste("Disruptive Inframe Insertion: ", length(grep("disruptive_inframe_insertion", variants$INFO))))
print(paste("Missense Variant: ", length(grep("missense_variant", variants$INFO))))
print(paste("Regulatory Region Ablation: ", length(grep("regulatory_region_ablation", variants$INFO))))
print(paste("Splice Region Variant: ", length(grep("splice_region_variant", variants$INFO))))
print(paste("TFBS Ablation: ", length(grep("TFBS_ablation", variants$INFO))))
print(paste("5 Prime UTR Premature Start Codon Gain Variant: ", length(grep("5_prime_UTR_premature_start_codon_gain_variant", variants$INFO))))
print(paste("Initiator Codon Variant: ", length(grep("initiator_codon_variant", variants$INFO))))
print(paste("Start Retained: ", length(grep("start_retained", variants$INFO))))
print(paste("Stop Retained Variant: ", length(grep("stop_retained_variant", variants$INFO))))
print(paste("Synonymous Variant: ", length(grep("synonymous_variant", variants$INFO))))
print(paste("3 Prime UTR Variant: ", length(grep("3_prime_UTR_variant", variants$INFO))))
print(paste("5 Prime UTR Variant: ", length(grep("5_prime_UTR_variant", variants$INFO))))
print(paste("Coding Sequence Variant: ", length(grep("coding_sequence_variant", variants$INFO))))
print(paste("Conserved Intergenic Variant: ", length(grep("conserved_intergenic variant", variants$INFO))))
print(paste("Conserved Intron Variant: ", length(grep("conserved_intron_variant", variants$INFO))))
print(paste("Downstream Gene Variant: ", length(grep("downstream_gene_variant", variants$INFO))))
print(paste("Exon Variant: ", length(grep("exon_variant", variants$INFO))))
print(paste("Feature Elongation: ", length(grep("feature_elongation", variants$INFO))))
print(paste("Feature Truncation: ", length(grep("feature_truncation", variants$INFO))))
print(paste("Gene Variant: ", length(grep("gene_variant", variants$INFO))))
print(paste("Intergenic Region: ", length(grep("intergenic_variant", variants$INFO))))
print(paste("Intragenic Variant: ", length(grep("intragenic_variant", variants$INFO))))
print(paste("Intron Variant: ", length(grep("intron_variant", variants$INFO))))
print(paste("Mature miRNA Variant: ", length(grep("mature_miRNA_variant", variants$INFO))))
print(paste("miRNA: ", length(grep("miRNA", variants$INFO))))
print(paste("NMD Transcript Variant: ", length(grep("NMD_transcript_variant", variants$INFO))))
print(paste("Non-Coding Transcript Exon Variant: ", length(grep("non_coding_transcript_exon_variant", variants$INFO))))
print(paste("Non-Coding Transcript Variant: ", length(grep("non_coding_transcript_variant", variants$INFO))))
print(paste("Regulatory Region Amplification: ", length(grep("regulatory_region_amplification", variants$INFO))))
print(paste("Regulatory Region Variant: ", length(grep("regulatory_region_variant", variants$INFO))))
print(paste("TF Binding Site Variant: ", length(grep("TF_binding_site_variant", variants$INFO))))
print(paste("TFBS Amplification: ", length(grep("TFBS_amplification", variants$INFO))))
print(paste("Transcript Amplification: ", length(grep("transcript_amplification", variants$INFO))))
print(paste("Transcript Variant: ", length(grep("transcript_variant", variants$INFO))))
print(paste("Upstream Gene Variant: ", length(grep("upstream_gene_variant", variants$INFO))))
```

Question 3.6
A frameshift variant is an insertion or deletion that involves a number of base pairs that is not 3, therefore it changes all of the codons that come after it. A missense variant results in the changing of a single nucleotide, therefore it changes only one codon and thus one amino. Therefore, frameshift variants are more harmful because they result in the creating of many wrong amino acids, compared to only one wrong amino acid in missense variations.

Question 3.7
```{r}
length(grep("intron",variants$INFO, value=FALSE))
```
The number of intronic variants is 567 out of 837 variants all together. This is a large amount of the overall variants.

Question 3.8 
```{r}
HighIndices<-grep("HIGH",variants$INFO, value=FALSE)
print("The variants with HIGH impact come from genes Ddx1, Rps14, Rps19, and Hnrnpl")


```
Question 3.9
Insertions that are longer than 60bp will be split up into multiple reads in the BAM file. Each read will then be aligned with the reference sequence, and since the insertion is longer than the read length, there is no way to identify it as an insertion. Insertions can only be seen if they are within the read length, because then when it is compared to the reference sequence, some portion of the sample sequence at the beginning and end will match the reference and a section in the middle can be identified as an insertion. 

Question 3.10
*Referenced from https://gatk.broadinstitute.org/hc/en-us/articles/360035531692-VCF-Variant-Call-Format
and https://stackoverflow.com/a/19002916/12129681
```{r}
library("vcfR")
# 
# #AD_frequency() function test:
# 
# set.seed(999)
# x1 <- round(rnorm(n=9, mean=10, sd=2))
# x2 <- round(rnorm(n=9, mean=20, sd=2))
# ad <- matrix(paste(x1, x2, sep=","), nrow=3, ncol=3)
# colnames(ad) <- paste('Sample', 1:3, sep="_")
# rownames(ad) <- paste('Variant', 1:3, sep="_")
# ad[1,1] <- "23,12"
# AD_frequency(ad, allele = 2, sum_type = 1)

#Make AD frequency matrix from split/matching AD entries in variants data frame
q310use <- unlist(mapply(function(A, B) unlist(strsplit(B, ":"))[match("AD", unlist(strsplit(A, ":")))], A=variants$FORMAT, B=variants$ws20171223_MPs_tomatoMuscle8wkQuiescent201))
q310split <- strsplit(q310use, ",")

q310alt <- as.numeric(unlist(lapply(q310split, function(s) unlist(strsplit(s, ","))[2])))
q310ref <- as.numeric(unlist(lapply(q310split, function(s) unlist(strsplit(s, ","))[1])))

q310boxplot <- q310alt / (q310ref + q310alt)

boxplot(q310boxplot, ylab = "VAF Values", main = "VAF Boxplot over all Variants")

#FIXME: Fixed issue, but before the AD_frequency function would take max(VAF, 1 - VAF) for some reason

over5percent <- which(q310boxplot > 0.05)
```
According to our processes, 823 variants have VAF > 5%.


